### PR TITLE
Settings: Reimplement keystore format to use FIPS compliant algorithms

### PR DESCRIPTION
**Pull request by rjernst**
This commit switches the internal format of the elasticsearch keystore
to no longer use java's KeyStore class, but instead encrypt the binary
data of the secrets using AES-GCM. The cipher key is generated using
PBKDF2WithHmacSHA512. Tests are also added for backcompat reading the v1
and v2 formats.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
I added `:Settings` and `v7.0.0` labels - please add more if needed.

**Review by jaymode — APPROVED** [review]
I left one comment, but other than that LGTM

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java [hunk 1]
```diff
@@ -210,71 +212,192 @@ public static KeyStoreWrapper load(Path configDir) throws IOException {
...
+                output.writeUTF(entry.type.name());
+                output.writeInt(entry.bytes.length);
+                output.write(entry.bytes);
+            }
+        }
+
+        return bytes.toByteArray();
+    }
+
+    private void decryptLegacyEntries() throws GeneralSecurityException, IOException {
```
Comment (jkakavas) [inline server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: This won't work if the JVM is in FIPS approved mode only as `PBE` is not available and ` SecretKeyFactory.getInstance("PBE");` will throw an Exception. 
I'm thinking of the case where someone wants to upgrade to 6.3.0 and at the same time enable FIPS approved only mode. We can highlight this and document manual steps that will ensure the keystore gets upgraded before FIPS approved mode is enabled.

**Review by jkakavas — APPROVED** [review]
LGTM, a minor comment on testing and a mental not for documentation

File: server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java (hunk 1 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: We can add an assumeFalse once we are ready to test that.

**Pull request merged**